### PR TITLE
libfabric: fix active rail tracking

### PR DESCRIPTION
**Pull request by fengjica**
Fix active rail tracking in libfabric backend.

## What?
Today there is no reference counting to correctly keep track of the status of rails. In addition to fixing that, I add a unit test with some code refactoring to avoid duplicating the common test code.

## Why?
It is a bug without this mechanism.

## Summary by CodeRabbit

* **New Features**
  * Active-rail tracking now uses per-rail reference counts; rails become inactive only when their count reaches zero.

* **Tests**
  * Added a comprehensive unit test covering refcount transitions, edge cases, memory-registration interactions, and multi-rail behavior.

* **Chores**
  * Added shared test stubs/helpers and a new test executable to support the refcounting tests.

### CODE REVIEW

File: src/utils/libfabric/libfabric_rail_manager.cpp [hunk 1]
```diff
@@ -741,7 +741,7 @@ nixlLibfabricRailManager::registerMemory(void *buffer,
         key_list_out[rail_idx] = key;
 
         // Mark rail as active for progress tracking optimization
-        markRailActive(rail_idx);
+        incRailActive(rail_idx);
```
[the remark this answers is not in the document]
Comment (fengjica) [reply src/utils/libfabric/libfabric_rail_manager.cpp:797]: Fixed.

File: test/unit/utils/libfabric/libfabric_mock_stubs.h [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+ */
+
+/*
+ * Shared libfabric stub ops for unit tests that need to construct a
+ * nixlLibfabricRailManager without real hardware.  Include this header
+ * in exactly one translation unit per test executable (it contains
+ * definitions, not just declarations).
+ */
+#ifndef NIXL_TEST_LIBFABRIC_MOCK_STUBS_H
+#define NIXL_TEST_LIBFABRIC_MOCK_STUBS_H
```
[the remark this answers is not in the document]
Comment (fengjica) [reply test/unit/utils/libfabric/libfabric_mock_stubs.h]: Fixed.

File: test/unit/utils/libfabric/rail_active_refcount_test.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+    nixlLibfabricRailManager mgr(0);
+    TEST_ASSERT(mgr.getNumRails() == NUM_FAKE_RAILS,
+                "expected " + std::to_string(NUM_FAKE_RAILS) + " rails");
+
+    int res;
+    if ((res = testIncDecBasic(mgr)) != 0) return res;
+    if ((res = testDecNonExistent(mgr)) != 0) return res;
+    if ((res = testIncInvalidRailId(mgr)) != 0) return res;
+    if ((res = testClearActiveRails(mgr)) != 0) return res;
+    if ((res = testMultipleRailsIndependent(mgr)) != 0) return res;
```
[the remark this answers is not in the document]
Comment (fengjica) [reply test/unit/utils/libfabric/rail_active_refcount_test.cpp:284]: fixed.

File: src/utils/libfabric/libfabric_rail_manager.cpp [hunk 4]
```diff
@@ -741,7 +743,7 @@ nixlLibfabricRailManager::registerMemory(void *buffer,
...
```
Comment (anshumang) [inline src/utils/libfabric/libfabric_rail_manager.cpp:796]: Since rail 0 is already unconditionally polled in `progressActiveRails()` and CQ completions don't distinguish 
control sends, do we need this `incRailActive()` here?

File: src/utils/libfabric/libfabric_rail_manager.cpp [hunk 5]
```diff
@@ -774,7 +776,7 @@ nixlLibfabricRailManager::deregisterMemory(const std::vector<size_t> &selected_r
                 NIXL_ERROR << "Failed to deregister memory on rail " << rail_idx;
                 overall_status = status;
             }
-            markRailInactive(rail_idx);
+            decRailActive(rail_idx);
```
Comment (anshumang) [inline src/utils/libfabric/libfabric_rail_manager.cpp]: If `deregisterMemory()` fails and the MR is still live, should we skip the `decRailActive()` here to avoid dropping the rail from the active set prematurely?

**Review by anshumang — CHANGES_REQUESTED** [review]

File: src/utils/libfabric/libfabric_rail_manager.cpp [hunk 6]
```diff
@@ -1131,30 +1135,31 @@ nixlLibfabricRailManager::deserializeRailEndpoints(
...
-nixlLibfabricRailManager::markRailInactive(size_t rail_id) {
+nixlLibfabricRailManager::decRailActive(size_t rail_id) {
     std::lock_guard<std::mutex> lock(active_rails_mutex_);
-    size_t erased = active_rails_.erase(rail_id);
-    if (erased > 0) {
-        NIXL_DEBUG << "Marked rail " << rail_id
-                   << " as inactive (total active: " << active_rails_.size() << ")";
+    auto it = active_rails_.find(rail_id);
+    if (it != active_rails_.end()) {
+        it->second--;
```
Comment (yexiang-aws) [inline src/utils/libfabric/libfabric_rail_manager.cpp]: nit: refcount`it->second` is size_t, if there's a bug over decrement it, it might underflow, do you want to add an assert `refcount==0` there before decrementing it?

File: src/utils/libfabric/libfabric_rail_manager.cpp [hunk 7]
```diff
@@ -714,6 +714,7 @@ nixlLibfabricRailManager::registerMemory(void *buffer,
                 if (mr_list_out[cleanup_idx]) {
                     rails_[cleanup_idx]->deregisterMemory(mr_list_out[cleanup_idx]);
                     mr_list_out[cleanup_idx] = nullptr;
+                    decRailActive(cleanup_idx);
```
Comment (yexiang-aws) [inline src/utils/libfabric/libfabric_rail_manager.cpp]: both here and line 736, we would need to check the status of `deregisterMemory` and if it failed, we should not decrement

File: src/utils/libfabric/libfabric_rail_manager.cpp (hunk 4 above)
Comment (fengjica) [reply src/utils/libfabric/libfabric_rail_manager.cpp:796]: After looking at the code, I see that another incRailActive(0) in postControlMessage() is actually unnecessary, and will indefinitely increment it in every postControlMessage(). Removed that one.

File: src/utils/libfabric/libfabric_rail_manager.cpp (hunk 5 above)
Comment (fengjica) [reply src/utils/libfabric/libfabric_rail_manager.cpp]: Makes sense.

Fixed in the next revision.

File: src/utils/libfabric/libfabric_rail_manager.cpp (hunk 6 above)
Comment (fengjica) [reply src/utils/libfabric/libfabric_rail_manager.cpp]: Fixed.

File: src/utils/libfabric/libfabric_rail_manager.cpp (hunk 7 above)
Comment (fengjica) [reply src/utils/libfabric/libfabric_rail_manager.cpp]: Make sense.

Fixed.

**Review by anshumang — APPROVED** [review]

**Review by yexiang-aws — APPROVED** [review]
